File: src/monitoring.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use chrono::{DateTime, Utc};

#[derive(Debug, Serialize, Deserialize)]
pub struct AlgorithmResult {
    pub algorithm_name: String,
    pub parameters: HashMap<String, String>,
    pub route_length: f64,
    pub route: Vec<usize>,
    pub execution_time_ms: u64,
    pub iterations: u32,
    pub start_timestamp: DateTime<Utc>,
    pub additional_metrics: HashMap<String, f64>,
}

impl AlgorithmResult {
    pub fn new(algorithm_name: String) -> Self {
        Self {
            algorithm_name,
            parameters: HashMap::new(),
            route_length: 0.0,
            route: Vec::new(),
            execution_time_ms: 0,
            iterations: 0,
            start_timestamp: Utc::now(),
            additional_metrics: HashMap::new(),
        }
    }

    pub fn with_parameter(mut self, key: String, value: String) -> Self {
        self.parameters.insert(key, value);
        self
    }

    pub fn with_metric(mut self, key: String, value: f64) -> Self {
        self.additional_metrics.insert(key, value);
        self
    }

    pub fn set_result(mut self, route_length: f64, route: Vec<usize>) -> Self {
        self.route_length = route_length;
        self.route = route;
        self
    }

    pub fn set_execution_time(mut self, time_ms: u64) -> Self {
        self.execution_time_ms = time_ms;
        self
    }

    pub fn set_iterations(mut self, iterations: u32) -> Self {
        self.iterations = iterations;
        self
    }
}

pub struct ResultMonitor {
    results_dir: String,
}

impl ResultMonitor {
    pub fn new() -> Self {
        let results_dir = "wyniki".to_string();
        Self { results_dir }
    }

    pub fn save_result(&self, result: &AlgorithmResult) -> Result<String, Box<dyn std::error::Error>> {
        // Ensure results directory exists
        if !Path::new(&self.results_dir).exists() {
            fs::create_dir_all(&self.results_dir)?;
        }

        // Create filename with algorithm name and timestamp
        let timestamp = result.start_timestamp.format("%Y%m%d_%H%M%S_%3f");
        let filename = format!("{}_{}.json", result.algorithm_name, timestamp);
        let filepath = Path::new(&self.results_dir).join(&filename);

        // Serialize to JSON
        let json = serde_json::to_string_pretty(result)?;

        // Write to file
        fs::write(&filepath, json)?;

        Ok(filename)
    }

    pub fn list_results(&self) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        if !Path::new(&self.results_dir).exists() {
            return Ok(Vec::new());
        }

        let mut files = Vec::new();
        for entry in fs::read_dir(&self.results_dir)? {
            let entry = entry?;
            if entry.path().extension().and_then(|s| s.to_str()) == Some("json") {
                if let Some(filename) = entry.file_name().to_str() {
                    files.push(filename.to_string());
                }
            }
        }
        files.sort();
        Ok(files)
    }
}

impl Default for ResultMonitor {
    fn default() -> Self {
        Self::new()
    }
}

File: src/utils/monitoring.py
"""
Moduł monitorowania wyników algorytmów TSP.

Ten moduł zapewnia kompleksowe śledzenie i analizę wyników algorytmów TSP.
"""

import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List

from utils.data_loader import TspDataset


@dataclass
class AlgorithmResult:
    """Reprezentuje wynik pojedynczego wykonania algorytmu."""

    algorithm_name: str
    parameters: Dict[str, str] = field(default_factory=dict)
    route_length: float = 0.0
    route: List[int] = field(default_factory=list)
    execution_time_ms: int = 0
    iterations: int = 0
    start_timestamp: datetime = field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    additional_metrics: Dict[str, float] = field(default_factory=dict)
    dataset_size: int = 0
    dataset_name: str = ""

    @classmethod
    def new(cls, algorithm_name: str) -> "AlgorithmResult":
        """Tworzy nowy AlgorithmResult z podaną nazwą algorytmu."""
        return cls(algorithm_name=algorithm_name)

    def with_dataset(self, dataset: TspDataset) -> "AlgorithmResult":
        """Ustawia informacje o zbiorze danych."""
        self.dataset_size = dataset.city_count()
        self.dataset_name = dataset.name()
        return self

    def with_parameter(self, key: str, value: str) -> "AlgorithmResult":
        """Dodaje parametr."""
        self.parameters[key] = value
        return self

    def with_metric(self, key: str, value: float) -> "AlgorithmResult":
        """Dodaje dodatkową metrykę."""
        self.additional_metrics[key] = value
        return self

    def set_result(self, route_length: float, route: List[int]) -> "AlgorithmResult":
        """Ustawia wynik (długość trasy i trasę)."""
        self.route_length = route_length
        self.route = route
        return self

    def set_execution_time(self, time_ms: int) -> "AlgorithmResult":
        """Ustawia czas wykonania w milisekundach."""
        self.execution_time_ms = time_ms
        return self

    def set_iterations(self, iterations: int) -> "AlgorithmResult":
        """Ustawia liczbę iteracji."""
        self.iterations = iterations
        return self

    def to_dict(self) -> dict:
        """Konwertuje do słownika do serializacji JSON."""
        return {
            "algorithm_name": self.algorithm_name,
            "parameters": self.parameters,
            "route_length": self.route_length,
            "route": self.route,
            "execution_time_ms": self.execution_time_ms,
            "iterations": self.iterations,
            "start_timestamp": self.start_timestamp.isoformat(),
            "additional_metrics": self.additional_metrics,
            "dataset_size": self.dataset_size,
            "dataset_name": self.dataset_name,
        }


class ResultMonitor:
    """Zarządza zbieraniem i trwałym przechowywaniem wyników algorytmów."""

    def __init__(self, results_dir: str = "wyniki"):
        """
        Inicjalizuje ResultMonitor.

        Args:
            results_dir: Katalog do przechowywania plików wyników
        """
        # Jeśli ścieżka nie jest bezwzględna, oblicz ją względem katalogu głównego projektu
        results_path = Path(results_dir)
        if not results_path.is_absolute():
            # Ścieżka do katalogu głównego projektu (poziom wyżej niż src/)
            project_root = Path(__file__).parent.parent.parent
            results_path = project_root / results_dir
        self.results_dir = str(results_path)

    def save_result(self, result: AlgorithmResult) -> str:
        """
        Zapisuje wynik algorytmu do pliku JSON.

        Args:
            result: AlgorithmResult do zapisania

        Returns:
            str: Nazwa pliku zapisanego wyniku

        Raises:
            IOError: Jeśli plik nie może zostać zapisany
        """
        # Upewnij się, że katalog wyników istnieje
        Path(self.results_dir).mkdir(parents=True, exist_ok=True)

        # Utwórz nazwę pliku z nazwą algorytmu, rozmiarem zbioru danych i znacznikiem czasu
        timestamp = result.start_timestamp.strftime("%Y%m%d_%H%M%S_%f")[
            :-3
        ]  # milisekundy
        filename = (
            f"{result.algorithm_name}_{result.dataset_size}cities_{timestamp}.json"
        )
        filepath = Path(self.results_dir) / filename

        # Konwertuj do słownika i serializuj do JSON
        result_dict = result.to_dict()
        json_str = json.dumps(result_dict, indent=2, ensure_ascii=False)

        # Zapisz do pliku
        with open(filepath, "w", encoding="utf-8") as f:
            f.write(json_str)

        return filename

    def list_results(self) -> List[str]:
        """
        Wyświetla listę wszystkich plików wyników w katalogu wyników.

        Returns:
            List[str]: Lista nazw plików wyników, posortowana alfabetycznie
        """
        results_path = Path(self.results_dir)

        if not results_path.exists():
            return []

        files = []
        for entry in results_path.iterdir():
            if entry.is_file() and entry.suffix == ".json":
                files.append(entry.name)

        files.sort()
        return files
